src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+            private long? _contentLengthRemaining;
+            private long _firstFrameLengthRemaining;
+            private byte[] _dataFrame;
+
+            public override bool CanRead => false;
+
+            public override bool CanWrite => true;
+
+            public long FirstFrameLengthRemaining => _firstFrameLengthRemaining;
+
+            public Http3WriteStream(QuicStream stream, long? contentLength, long firstFrameLengthRemaining)
+            {
+                _stream = stream;
+                _firstFrameLengthRemaining = firstFrameLengthRemaining;
+                _contentLengthRemaining = contentLength;
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                _stream = null;
+                base.Dispose(disposing);
+            }
+
+            public override int Read(Span<byte> buffer)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            {
+                QuicStream stream = _stream ?? throw new ObjectDisposedException(nameof(Http3WriteStream));
+
+                if (buffer.Length == 0)
+                {
+                    return;
+                }
+
+                if (buffer.Length > _contentLengthRemaining)
+                {
+                    string msg = "Unable to write content larger than what was specified in Content-Length header.";
+                    throw new IOException(msg, new HttpRequestException(msg));
+                }
+                _contentLengthRemaining -= buffer.Length;
+
+                long remaining = _firstFrameLengthRemaining;
+                if (remaining > 0)
+                {
+                    // This HttpContent had a precomputed length, and a data frame was written as part of the headers. We're writing against the first frame.
+
+                    int sendLength = (int)Math.Min(remaining, buffer.Length);
+                    await stream.WriteAsync(buffer.Slice(0, sendLength), cancellationToken).ConfigureAwait(false);
+
+                    buffer = buffer.Slice(sendLength);
+                    _firstFrameLengthRemaining -= sendLength;
+
+                    if (buffer.Length == 0)
+                    {
+                        return;
+                    }
+                }
+
+                // This is only reached if the HttpContent couldn't precompute its length, or if it
+                // writes more bytes than was precomputed (this latter case is probably an error -- TODO).
+                // In this case, each write results in a data frame. TODO: optimize this to avoid small frame sizes.
+
+                byte[] dataFrame = _dataFrame;
```

Comment:
```suggestion
                byte[] dataFrame = _dataFrame ??= new byte[9];
```
Saves some C# code when the following `if` is removed. More or less a matter of taste.

Reply:
I'm fine with it now; will update if there are strong opinions.